src/main/scala/com/microsoft/hyperspace/index/rules/FilterIndexRule.scala

```diff
@@ -106,29 +107,56 @@ object FilterIndexRule
         val newLocation =
           new InMemoryFileIndex(spark, Seq(new Path(index.content.root)), Map(), None)
 
-        val newRelation = HadoopFsRelation(
+        val newFsRelation = HadoopFsRelation(
           newLocation,
           new StructType(),
           index.schema,
-          None, // Do not set BucketSpec to avoid limiting Spark's degree of parallelism
+          Some(index.bucketSpec),
           new ParquetFileFormat,
           Map())(spark)
 
         val newOutput =
           logicalRelation.output.filter(attr => index.schema.fieldNames.contains(attr.name))
 
-        val updatedPlan =
-          filter.copy(child = logicalRelation.copy(relation = newRelation, output = newOutput))
+        val newIndexRelation = logicalRelation.copy(relation = newFsRelation, output = newOutput)
+        val updatedPlan = filter.copy(child = newIndexRelation)
+
+        // for appended source files
+        val appendedSourceFiles =
+          fsRelation.location.inputFiles.map(new Path(_))
+            .filter(p => !index.allSourceFileSet.contains(p))
+
+        val replacedPlan = if (!appendedSourceFiles.isEmpty) {
+          val appendedLocation =
+            new InMemoryFileIndex(spark, appendedSourceFiles, Map(), None)
+          val appendedRelation = logicalRelation.copy(
+            relation =
+              fsRelation.copy(location = appendedLocation, dataSchema = index.schema)(spark),
+            output = newOutput)
+
+          val attrs = appendedRelation.attributeMap.baseMap
+            .filter { key =>
+              index.indexedColumns contains key._2._2.name
+            }
+            .map { case (_, ref) => ref._1 }
+
+          val filtered = filter.copy(child = appendedRelation)
+          val shuffled =
+            RepartitionByExpression(attrs.toSeq, filtered, index.numBuckets)
+          // val filtered2 = filter.copy(child = shuffled)
+          BucketUnionLogicalPlan(Seq(updatedPlan, shuffled), index.bucketSpec)
```

Comment:
should we remove sort order from bucketSpec since only index is sorted, data isn't?

Reply:
cc @imback82